File: redemption.py
"""RedemptionMixin: on-chain CTF position redemption via Polymarket builder relayer."""

from __future__ import annotations

import os
import time as _time
from datetime import datetime, timezone
from typing import TYPE_CHECKING

import requests
from web3 import Web3
from eth_abi import encode as abi_encode
from eth_account import Account as EthAccount
from eth_account.messages import encode_defunct
from py_builder_signing_sdk.signing.hmac import build_hmac_signature

if TYPE_CHECKING:
    from tracker import LiveTradeTracker

# ── On-chain redemption constants ────────────────────────────────────────────
POLYGON_RPC = os.getenv("POLYGON_RPC", "https://polygon-rpc.com")

CTF_ADDRESS = Web3.to_checksum_address("0x4D97DCd97eC945f40cF65F87097ACe5EA0476045")
USDC_ADDRESS = Web3.to_checksum_address("0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174")

CTF_ABI = [
    {
        "inputs": [
            {"name": "collateralToken", "type": "address"},
            {"name": "parentCollectionId", "type": "bytes32"},
            {"name": "conditionId", "type": "bytes32"},
            {"name": "indexSets", "type": "uint256[]"},
        ],
        "name": "redeemPositions",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
    },
    {
        "inputs": [{"name": "conditionId", "type": "bytes32"}],
        "name": "payoutDenominator",
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function",
    },
]

PARENT_COLLECTION_ID = bytes(32)  # 0x0...0

# ── Builder relayer constants (PROXY type) ──────────────────────────────────
RELAYER_URL = "https://relayer-v2.polymarket.com"
PROXY_FACTORY = Web3.to_checksum_address("0xaB45c5A4B0c941a2F231C04C3f49182e1A254052")
RELAY_HUB = Web3.to_checksum_address("0xD216153c06E857cD7f72665E0aF1d7D82172F494")
PROXY_INIT_CODE_HASH = bytes.fromhex(
    "d21df8dc65880a8606f09fe0ce3df9b8869287ab0b058be05aa9e8af6330a00b"
)
DEFAULT_RELAY_GAS_LIMIT = 500_000


class RedemptionMixin:
    """On-chain redemption methods mixed into LiveTradeTracker."""

    def redeem_positions(
        self: "LiveTradeTracker",
        condition_id: str,
        max_retries: int = 3,
        max_poll_attempts: int = 60,
        poll_interval_s: float = 30.0,
    ) -> str | None:
        """Redeem winning CTF positions via the Polymarket builder relayer.

        First polls payoutDenominator on-chain to wait for resolution,
        then submits via the builder relayer (no POL gas needed).
        Returns the tx hash on success, None on failure.
        """
        if self.dry_run:
            print("  [REDEEM] Skipped (dry run)")
            return None

        if not condition_id:
            print("  [REDEEM] Skipped (no conditionId)")
            return None

        # Poll payoutDenominator until on-chain resolution arrives
        condition_bytes = bytes.fromhex(condition_id.replace("0x", ""))
        resolved = False
        try:
            w3 = Web3(Web3.HTTPProvider(POLYGON_RPC))
            if not w3.is_connected():
                print("  [REDEEM] ERROR: Cannot connect to Polygon RPC")
                return None
            ctf = w3.eth.contract(address=CTF_ADDRESS, abi=CTF_ABI)
            for poll in range(max_poll_attempts):
                try:
                    denom = ctf.functions.payoutDenominator(condition_bytes).call()
                    if denom > 0:
                        resolved = True
                        print(f"  [REDEEM] On-chain resolution confirmed "
                              f"(payoutDenominator={denom}, poll #{poll + 1})")
                        break
                except Exception as exc:
                    if self.debug:
                        print(f"  [REDEEM] payoutDenominator poll error: {exc}")
                if poll < max_poll_attempts - 1:
                    if poll == 0:
                        print(f"  [REDEEM] Waiting for on-chain resolution "
                              f"(polling every {poll_interval_s:.0f}s, "
                              f"up to {max_poll_attempts * poll_interval_s / 60:.0f}m)...")
                    _time.sleep(poll_interval_s)
        except Exception as exc:
            print(f"  [REDEEM] RPC connection error: {exc}")

        if not resolved:
            print(f"  [REDEEM] On-chain resolution not found after "
                  f"{max_poll_attempts * poll_interval_s / 60:.0f}m. "
                  f"Claim manually on Polymarket.")
            return None

        for attempt in range(max_retries):
            if attempt > 0:
                delay = 20 * attempt
                print(f"  [REDEEM] Retry {attempt + 1}/{max_retries} in {delay}s...")
                _time.sleep(delay)

            result = self._try_redeem_once(condition_id)
            if result is not None:
                return result

        print(f"  [REDEEM] Failed after {max_retries} tx attempts. "
              f"Claim manually on Polymarket.")
        return None

    def _try_redeem_once(self: "LiveTradeTracker", condition_id: str) -> str | None:
        """Single attempt at CTF redemption via Polymarket builder relayer (PROXY type)."""
        try:
            private_key = os.getenv("PRIVATE_KEY", "")
            if not private_key:
                print("  [REDEEM] ERROR: PRIVATE_KEY not set")
                return None
            if not private_key.startswith("0x"):
                private_key = "0x" + private_key

            builder_key = os.getenv("POLY_BUILDER_API_KEY", "")
            builder_secret = os.getenv("POLY_BUILDER_SECRET", "")
            builder_passphrase = os.getenv("POLY_BUILDER_PASSPHRASE", "")
            if not all([builder_key, builder_secret, builder_passphrase]):
                print("  [REDEEM] ERROR: Builder API creds not set "
                      "(POLY_BUILDER_API_KEY, POLY_BUILDER_SECRET, "
                      "POLY_BUILDER_PASSPHRASE)")
                return None

            signer = EthAccount.from_key(private_key)

            # 1. Encode CTF.redeemPositions(USDC, 0x0, conditionId, [1, 2])
            w3 = Web3()
            ctf = w3.eth.contract(address=CTF_ADDRESS, abi=CTF_ABI)
            condition_bytes = bytes.fromhex(condition_id.replace("0x", ""))
            redeem_data = ctf.encode_abi(
                "redeemPositions",
                args=[USDC_ADDRESS, PARENT_COLLECTION_ID, condition_bytes, [1, 2]],
            )
            if isinstance(redeem_data, str):
                redeem_bytes = bytes.fromhex(
                    redeem_data[2:] if redeem_data.startswith("0x") else redeem_data
                )
            else:
                redeem_bytes = bytes(redeem_data)

            # 2. Wrap in proxy((uint8,address,uint256,bytes)[])
            proxy_selector = w3.keccak(
                b"proxy((uint8,address,uint256,bytes)[])"
            )[:4]
            proxy_args = abi_encode(
                ["(uint8,address,uint256,bytes)[]"],
                [[(1, CTF_ADDRESS, 0, redeem_bytes)]],
            )
            proxy_data = "0x" + (proxy_selector + proxy_args).hex()

            # 3. Derive proxy wallet via CREATE2
            salt = w3.keccak(bytes.fromhex(signer.address[2:].lower()))
            create2_input = (
                b"\xff"
                + bytes.fromhex(PROXY_FACTORY[2:])
                + salt
                + PROXY_INIT_CODE_HASH
            )
            proxy_wallet = Web3.to_checksum_address(
                "0x" + w3.keccak(create2_input).hex()[-40:]
            )

            # 4. Get relay payload (nonce + relay address)
            headers = self._builder_headers(
                builder_key, builder_secret, builder_passphrase,
                "GET", f"/relay-payload?address={signer.address}&type=PROXY",
            )
            resp = requests.get(
                f"{RELAYER_URL}/relay-payload",
                params={"address": signer.address, "type": "PROXY"},
                headers=headers,
            )
            resp.raise_for_status()
            payload = resp.json()
            relay_address = payload["address"]
            nonce = payload["nonce"]

            # 5. Create struct hash and sign (EIP-191)
            gas_limit = DEFAULT_RELAY_GAS_LIMIT
            struct_data = (
                b"rlx:"
                + bytes.fromhex(signer.address[2:])
                + bytes.fromhex(PROXY_FACTORY[2:])
                + bytes.fromhex(proxy_data[2:])
                + int(0).to_bytes(32, "big")
                + int(0).to_bytes(32, "big")
                + int(gas_limit).to_bytes(32, "big")
                + int(nonce).to_bytes(32, "big")
                + bytes.fromhex(RELAY_HUB[2:])
                + bytes.fromhex(relay_address.replace("0x", ""))
            )
            struct_hash = w3.keccak(struct_data)
            msg = encode_defunct(struct_hash)
            sig = EthAccount.sign_message(msg, private_key)
            signature = "0x" + sig.signature.hex()

            # 6. POST /submit
            body = {
                "type": "PROXY",
                "from": signer.address,
                "to": PROXY_FACTORY,
                "proxyWallet": proxy_wallet,
                "data": proxy_data,
                "nonce": str(nonce),
                "signature": signature,
                "signatureParams": {
                    "gasPrice": "0",
                    "gasLimit": str(gas_limit),
                    "relayerFee": "0",
                    "relayHub": RELAY_HUB,
                    "relay": relay_address,
                },
                "metadata": "Redeem winnings",
            }

            print(f"  [REDEEM] Submitting via relayer "
                  f"(conditionId={condition_id[:10]}..., "
                  f"proxy={proxy_wallet[:10]}...)...")

            submit_headers = self._builder_headers(
                builder_key, builder_secret, builder_passphrase,
                "POST", "/submit", body,
            )
            submit_headers["Content-Type"] = "application/json"
            resp = requests.post(
                f"{RELAYER_URL}/submit",
                json=body,
                headers=submit_headers,
            )
            resp.raise_for_status()
            result = resp.json()
            if self.debug:
                print(f"  [REDEEM] Relayer response: {result}")
            tx_id = result.get("transactionID") or result.get("transactionId", "")
            tx_hash = result.get("transactionHash", "")

            print(f"  [REDEEM] Relayer accepted: id={tx_id}, hash={tx_hash}")

            if not tx_id:
                print(f"  [REDEEM] WARNING: Relayer returned no transaction ID. "
                      f"Raw response: {result}")
                return tx_hash if tx_hash else None

            print(f"  [REDEEM] Waiting for confirmation...")

            # 7. Poll /transaction until confirmed
            confirmed = False
            for poll in range(30):
                _time.sleep(2)
                poll_headers = self._builder_headers(
                    builder_key, builder_secret, builder_passphrase,
                    "GET", f"/transaction?id={tx_id}",
                )
                try:
                    r = requests.get(
                        f"{RELAYER_URL}/transaction",
                        params={"id": tx_id},
                        headers=poll_headers,
                    )
                    r.raise_for_status()
                    tx_data = r.json()
                    state = ""
                    if isinstance(tx_data, list) and tx_data:
                        state = tx_data[0].get("state", "")
                        tx_hash = tx_data[0].get("transactionHash", tx_hash)
                    elif isinstance(tx_data, dict):
                        state = tx_data.get("state", "")
                        tx_hash = tx_data.get("transactionHash", tx_hash)

                    if state in ("STATE_MINED", "STATE_CONFIRMED"):
                        confirmed = True
                        print(f"  [REDEEM] Confirmed! (state={state}, "
                              f"hash={tx_hash})")
                        break
                    elif state == "STATE_FAILED":
                        print(f"  [REDEEM] Relayer tx FAILED (id={tx_id})")
                        break
                except Exception as poll_exc:
                    if self.debug:
                        print(f"  [REDEEM] Poll error: {poll_exc}")

            if confirmed:
                self._log({
                    "type": "redemption",
                    "ts": datetime.now(timezone.utc).isoformat(),
                    "tx_hash": tx_hash,
                    "tx_id": tx_id,
                    "condition_id": condition_id,
                    "status": "success",
                })
                return tx_hash
            else:
                print(f"  [REDEEM] Relayer tx not confirmed after 60s "
                      f"(id={tx_id}, hash={tx_hash})")
                self._log({
                    "type": "redemption",
                    "ts": datetime.now(timezone.utc).isoformat(),
                    "tx_hash": tx_hash,
                    "tx_id": tx_id,
                    "condition_id": condition_id,
                    "status": "relayer_timeout",
                })
                return None

        except Exception as exc:
            print(f"  [REDEEM] ERROR: {type(exc).__name__}: {exc}")
            self._log({
                "type": "redemption",
                "ts": datetime.now(timezone.utc).isoformat(),
                "condition_id": condition_id,
                "status": "error",
                "error": str(exc),
            })
            return None

    @staticmethod
    def _builder_headers(key: str, secret: str, passphrase: str,
                         method: str, path: str, body=None) -> dict:
        """Generate HMAC auth headers for the Polymarket builder relayer."""
        ts = str(int(_time.time()))
        sig = build_hmac_signature(secret, ts, method, path, body)
        return {
            "POLY_BUILDER_API_KEY": key,
            "POLY_BUILDER_PASSPHRASE": passphrase,
            "POLY_BUILDER_SIGNATURE": sig,
            "POLY_BUILDER_TIMESTAMP": ts,
        }

File: latency-arb/rust/src/redemption.rs

use anyhow::{anyhow, bail, Context, Result};
use base64::{
    engine::general_purpose::{URL_SAFE, URL_SAFE_NO_PAD},
    Engine,
};
use hmac::{Hmac, Mac};
use sha2::Sha256;
use std::time::{SystemTime, UNIX_EPOCH};

use alloy_primitives::{address, hex, keccak256, Address, B256, U256};
use alloy_signer::SignerSync;
use alloy_signer_local::PrivateKeySigner;
use alloy_sol_types::{sol, SolCall};

const CTF_ADDRESS: Address = address!("0x4D97DCd97eC945f40cF65F87097ACe5EA0476045");
const USDC_ADDRESS: Address = address!("0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174");
const PROXY_FACTORY: Address = address!("0xaB45c5A4B0c941a2F231C04C3f49182e1A254052");
const RELAY_HUB: Address = address!("0xD216153c06E857cD7f72665E0aF1d7D82172F494");

const RELAYER_URL: &str = "https://relayer-v2.polymarket.com";

const PROXY_INIT_CODE_HASH: [u8; 32] = hex!(
    "d21df8dc65880a8606f09fe0ce3df9b8869287ab0b058be05aa9e8af6330a00b"
);

const DEFAULT_RELAY_GAS_LIMIT: u64 = 500_000;

sol! {
    function redeemPositions(
        address collateralToken,
        bytes32 parentCollectionId,
        bytes32 conditionId,
        uint256[] indexSets
    );

    struct ProxyArg {
        uint8 kind;
        address to;
        uint256 value;
        bytes data;
    }
    function proxy(ProxyArg[] txns);
}

fn hmac_sign(
    secret_b64: &str,
    timestamp: &str,
    method: &str,
    path: &str,
    body: Option<&str>,
) -> Result<String> {
    let trimmed = secret_b64.trim_end_matches('=');
    let secret = URL_SAFE_NO_PAD
        .decode(trimmed)
        .context("invalid builder secret base64")?;
    let mut msg = format!("{timestamp}{method}{path}");
    if let Some(b) = body {
        msg.push_str(b);
    }
    let mut mac = Hmac::<Sha256>::new_from_slice(&secret)
        .context("hmac key init")?;
    mac.update(msg.as_bytes());
    let digest = mac.finalize().into_bytes();
    Ok(URL_SAFE.encode(digest))
}

fn now_ts() -> String {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs().to_string())
        .unwrap_or_else(|_| "0".to_string())
}

fn header_set(
    req: reqwest::RequestBuilder,
    key: &str,
    passphrase: &str,
    sig: &str,
    ts: &str,
) -> reqwest::RequestBuilder {
    req.header("POLY_BUILDER_API_KEY", key)
        .header("POLY_BUILDER_PASSPHRASE", passphrase)
        .header("POLY_BUILDER_SIGNATURE", sig)
        .header("POLY_BUILDER_TIMESTAMP", ts)
}

fn derive_proxy_wallet(eoa: Address) -> Address {
    let salt: B256 = keccak256(eoa.as_slice());

    let mut create2_input = Vec::with_capacity(1 + 20 + 32 + 32);
    create2_input.push(0xff);
    create2_input.extend_from_slice(PROXY_FACTORY.as_slice());
    create2_input.extend_from_slice(salt.as_slice());
    create2_input.extend_from_slice(&PROXY_INIT_CODE_HASH);

    let hash = keccak256(&create2_input);
    let mut addr = [0u8; 20];
    addr.copy_from_slice(&hash[12..]);
    Address::from(addr)
}

fn struct_hash(
    signer_addr: Address,
    proxy_data: &[u8],
    gas_price: u64,
    relayer_fee: u64,
    gas_limit: u64,
    nonce: u64,
    relay_addr: Address,
) -> B256 {
    let mut buf = Vec::with_capacity(4 + 20 + 20 + proxy_data.len() + 32 * 4 + 20 + 20);
    buf.extend_from_slice(b"rlx:");
    buf.extend_from_slice(signer_addr.as_slice());
    buf.extend_from_slice(PROXY_FACTORY.as_slice());
    buf.extend_from_slice(proxy_data);
    buf.extend_from_slice(&U256::from(gas_price).to_be_bytes::<32>());
    buf.extend_from_slice(&U256::from(relayer_fee).to_be_bytes::<32>());
    buf.extend_from_slice(&U256::from(gas_limit).to_be_bytes::<32>());
    buf.extend_from_slice(&U256::from(nonce).to_be_bytes::<32>());
    buf.extend_from_slice(RELAY_HUB.as_slice());
    buf.extend_from_slice(relay_addr.as_slice());
    keccak256(&buf)
}

fn encode_redeem_calldata(condition_id: &str) -> Result<Vec<u8>> {
    let cid_hex = condition_id.trim_start_matches("0x");
    if cid_hex.len() != 64 {
        bail!("conditionId must be 32 hex bytes, got {}", cid_hex.len());
    }
    let cid_bytes = hex::decode(cid_hex).context("decode conditionId")?;
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&cid_bytes);

    let call = redeemPositionsCall {
        collateralToken: USDC_ADDRESS,
        parentCollectionId: B256::ZERO,
        conditionId: arr.into(),
        indexSets: vec![U256::from(1u64), U256::from(2u64)],
    };
    Ok(call.abi_encode())
}

fn encode_proxy_calldata(redeem_datas: Vec<Vec<u8>>) -> Vec<u8> {
    let txns: Vec<ProxyArg> = redeem_datas
        .into_iter()
        .map(|data| ProxyArg {
            kind: 1, // DelegateCall semantic per Polymarket proxy
            to: CTF_ADDRESS,
            value: U256::ZERO,
            data: data.into(),
        })
        .collect();
    let call = proxyCall { txns };
    call.abi_encode()
}

#[allow(dead_code)]
pub async fn redeem_position(http: &reqwest::Client, condition_id: &str) -> Result<String> {
    redeem_positions(http, std::slice::from_ref(&condition_id.to_string())).await
}

pub async fn redeem_positions(
    http: &reqwest::Client,
    condition_ids: &[String],
) -> Result<String> {
    if condition_ids.is_empty() {
        return Err(anyhow!("empty batch"));
    }
    if condition_ids.len() > 15 {
        return Err(anyhow!("max 15 per batch, got {}", condition_ids.len()));
    }
    let private_key = std::env::var("PRIVATE_KEY")
        .context("PRIVATE_KEY not set")?;
    let builder_key = std::env::var("POLY_BUILDER_API_KEY")
        .context("POLY_BUILDER_API_KEY not set")?;
    let builder_secret = std::env::var("POLY_BUILDER_SECRET")
        .context("POLY_BUILDER_SECRET not set")?;
    let builder_pass = std::env::var("POLY_BUILDER_PASSPHRASE")
        .context("POLY_BUILDER_PASSPHRASE not set")?;

    let pk_clean = private_key.trim_start_matches("0x");
    let signer: PrivateKeySigner = pk_clean.parse()
        .context("parse PRIVATE_KEY")?;
    let signer_addr = signer.address();
    let proxy_wallet = derive_proxy_wallet(signer_addr);

    let mut redeem_datas = Vec::with_capacity(condition_ids.len());
    for cid in condition_ids {
        redeem_datas.push(encode_redeem_calldata(cid)?);
    }
    let proxy_data = encode_proxy_calldata(redeem_datas);
    let proxy_data_hex = format!("0x{}", hex::encode(&proxy_data));

    // 3. GET /relay-payload to get nonce + relay address
    let payload_path = format!("/relay-payload?address={signer_addr}&type=PROXY");
    let ts = now_ts();
    let sig = hmac_sign(&builder_secret, &ts, "GET", &payload_path, None)?;
    let resp = header_set(
        http.get(format!("{RELAYER_URL}{payload_path}")),
        &builder_key,
        &builder_pass,
        &sig,
        &ts,
    )
    .send()
    .await
    .context("relay-payload request")?;
    let status = resp.status();
    if !status.is_success() {
        let body = resp.text().await.unwrap_or_default();
        bail!("relay-payload HTTP {status}: {body}");
    }
    let payload: serde_json::Value = resp.json().await.context("parse relay-payload")?;
    let relay_address_str = payload
        .get("address")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow!("relay-payload missing 'address'"))?
        .to_string();
    let relay_address: Address = relay_address_str
        .parse()
        .context("parse relay address")?;
    let nonce: u64 = payload
        .get("nonce")
        .and_then(|v| v.as_u64().or_else(|| v.as_str().and_then(|s| s.parse().ok())))
        .ok_or_else(|| anyhow!("relay-payload missing 'nonce'"))?;

    // 4. Sign the struct hash with EIP-191 personal_sign
    let hash = struct_hash(
        signer_addr,
        &proxy_data,
        0,
        0,
        DEFAULT_RELAY_GAS_LIMIT,
        nonce,
        relay_address,
    );
    let sig_raw = signer
        .sign_hash_sync(&hash)
        .context("sign struct hash")?;
    let mut eip191 = Vec::with_capacity(32 + 28);
    eip191.extend_from_slice(b"\x19Ethereum Signed Message:\n32");
    eip191.extend_from_slice(hash.as_slice());
    let eip191_hash = keccak256(&eip191);
    let signature = signer
        .sign_hash_sync(&eip191_hash)
        .context("sign eip191 hash")?;
    let sig_bytes = signature.as_bytes();
    let signature_hex = format!("0x{}", hex::encode(sig_bytes));
    let _ = sig_raw;

    let body_str = format!(
        concat!(
            r#"{{"type": "PROXY", "from": "{from}", "to": "{to}", "proxyWallet": "{proxy}", "#,
            r#""data": "{data}", "nonce": "{nonce}", "signature": "{sig}", "#,
            r#""signatureParams": {{"gasPrice": "0", "gasLimit": "{gas}", "relayerFee": "0", "#,
            r#""relayHub": "{hub}", "relay": "{relay}"}}, "metadata": "Redeem winnings"}}"#
        ),
        from = signer_addr,
        to = PROXY_FACTORY,
        proxy = proxy_wallet,
        data = proxy_data_hex,
        nonce = nonce,
        sig = signature_hex,
        gas = DEFAULT_RELAY_GAS_LIMIT,
        hub = RELAY_HUB,
        relay = relay_address_str, // raw casing from Polymarket
    );

    let ts2 = now_ts();
    let sig2 = hmac_sign(&builder_secret, &ts2, "POST", "/submit", Some(&body_str))?;

    let submit = header_set(
        http.post(format!("{RELAYER_URL}/submit")),
        &builder_key,
        &builder_pass,
        &sig2,
        &ts2,
    )
    .header("Content-Type", "application/json")
    .body(body_str)
    .send()
    .await
    .context("submit request")?;

    let submit_status = submit.status();
    if !submit_status.is_success() {
        let text = submit.text().await.unwrap_or_default();
        bail!("submit HTTP {submit_status}: {text}");
    }
    let submit_body: serde_json::Value = submit.json().await.context("parse submit")?;
    let tx_id = submit_body
        .get("transactionID")
        .or_else(|| submit_body.get("transactionId"))
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let mut tx_hash = submit_body
        .get("transactionHash")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    if tx_id.is_empty() {
        return Ok(tx_hash); // no tx id to poll; caller will verify on-chain later
    }

    // 6. Poll /transaction until state is terminal (~up to 2 min)
    for _ in 0..60 {
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        let poll_path = format!("/transaction?id={tx_id}");
        let ts3 = now_ts();
        let sig3 = hmac_sign(&builder_secret, &ts3, "GET", &poll_path, None)?;
        let poll = header_set(
            http.get(format!("{RELAYER_URL}{poll_path}")),
            &builder_key,
            &builder_pass,
            &sig3,
            &ts3,
        )
        .send()
        .await;
        let Ok(poll_resp) = poll else { continue };
        if !poll_resp.status().is_success() {
            continue;
        }
        let Ok(tx_data) = poll_resp.json::<serde_json::Value>().await else { continue };
        let (state, hash_from_poll) = if let Some(first) = tx_data.as_array().and_then(|a| a.first()) {
            (
                first.get("state").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                first
                    .get("transactionHash")
                    .and_then(|v| v.as_str())
                    .unwrap_or(&tx_hash)
                    .to_string(),
            )
        } else {
            (
                tx_data.get("state").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                tx_data
                    .get("transactionHash")
                    .and_then(|v| v.as_str())
                    .unwrap_or(&tx_hash)
                    .to_string(),
            )
        };
        if !hash_from_poll.is_empty() {
            tx_hash = hash_from_poll;
        }
        match state.as_str() {
            "STATE_MINED" | "STATE_CONFIRMED" => return Ok(tx_hash),
            "STATE_FAILED" => bail!("relayer tx FAILED id={tx_id} hash={tx_hash}"),
            _ => continue,
        }
    }
    bail!("relayer tx not confirmed after 120s id={tx_id} hash={tx_hash}")
}
